### PR TITLE
server: check if there are active nics before network GC

### LINKED ISSUE #8200: Isolated network is shutdown even when there are running VMs on it
in some cases, the isolated network is shutdown even when there running VMs on it. due to it
- the VR is stopped
- the vlan is released

As a workaround, to avoid the issue, user has to update the network to persistent, t

update network_offerings set is_persistent = 1 where traffic_type='Guest' and guest_type='Isolated';    

The issue is that, the nics_count in op_networks table is changed to 0, which is wrong sometimes.
We should add check on running/migrating/stopping VMs as well, before shutting down an isolated network, which is similar as
https://github.com/apache/cloudstack/blob/3d7281d4516e4a73a4a30921b9fe58a987fcb038/server/src/main/java/com/cloud/network/NetworkModelImpl.java#L2564-L2567

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~

~~~

##### CLOUDSTACK VERSION

~~~
4.18
~~~

##### STEPS TO REPRODUCE

~~~

~~~

##### EXPECTED RESULTS

~~~

~~~

##### ACTUAL RESULTS

~~~

~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #8200 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java [hunk 1]
```diff
@@ -2561,7 +2561,7 @@ public boolean isNetworkReadyForGc(long networkId) {
 
         //if the network has vms in Starting state (nics for those might not be allocated yet as Starting state also used when vm is being Created)
```
[the remark this answers is not in the document]
Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:2562]: Indeed,  thanks Hari. 
I will update this PR

**Comment by DaanHoogland** [comment]
Code looks good @weizhouapache , one concern about the situation the a VM is being created but does not have any nics yet. Is this the situation you describe in #8200? If GC hits at that time, it might be possible that the network is going to be teared down, while the process creating the VM thinks it can continue.
Am I right?

**Comment by weizhouapache** [comment]
> Code looks good @weizhouapache , one concern about the situation the a VM is being created but does not have any nics yet. 

This situlation has already been handled.

> Is this the situation you describe in #8200? If GC hits at that time, it might be possible that the network is going to be teared down, while the process creating the VM thinks it can continue. Am I right?

This PR aims to handle the situation that nics_count is 0 but there are running /migrating/stopping vms. Ths nics_count should not be 0 but we have seens this issue sometimes already. The root cause is still unknown. To prevent the network to be shutdown and VR to be stopped , we should consider the situation that nics_count is wrong as 0.

**Comment by DaanHoogland** [comment]
> > Code looks good @weizhouapache , one concern about the situation the a VM is being created but does not have any nics yet.
> 
> This situlation has already been handled.

How has it?

> > Is this the situation you describe in #8200? If GC hits at that time, it might be possible that the network is going to be teared down, while the process creating the VM thinks it can continue. Am I right?
> 
> This PR aims to handle the situation that nics_count is 0 but there are running /migrating/stopping vms. Ths nics_count should not be 0 but we have seens this issue sometimes already. The root cause is still unknown. To prevent the network to be shutdown and VR to be stopped , we should consider the situation that nics_count is wrong as 0.

I understand this part and I approve of this PR because of that.

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
Is there a reason this needs to stay in draft @weizhouapache ?

**Comment by weizhouapache** [comment]
> > > Code looks good @weizhouapache , one concern about the situation the a VM is being created but does not have any nics yet.
> > 
> > This situlation has already been handled.
> 
> How has it?
> 

@DaanHoogland 
When vm is being created,  there might be two vm state:  starting and stopped.  For stopped vms,  the nic is not considered active and it is not impacted if network is shutdown.  For starting vms,  there is already a check :  


        if (_nicDao.countNicsForStartingVms(networkId) > 0) {


> > > Is this the situation you describe in #8200? If GC hits at that time, it might be possible that the network is going to be teared down, while the process creating the VM thinks it can continue. Am I right?
> > 
> > This PR aims to handle the situation that nics_count is 0 but there are running /migrating/stopping vms. Ths nics_count should not be 0 but we have seens this issue sometimes already. The root cause is still unknown. To prevent the network to be shutdown and VR to be stopped , we should consider the situation that nics_count is wrong as 0.
> 
> I understand this part and I approve of this PR because of that.

**Comment by DaanHoogland** [comment]
> @DaanHoogland When vm is being created, there might be two vm state: starting and stopped. For stopped vms, the nic is not considered active and it is not impacted if network is shutdown. For starting vms, there is already a check :
> 
> ```
>     if (_nicDao.countNicsForStartingVms(networkId) > 0) {
> ```

Getting into "muggenzift"a area here but,
are we sure the nic is created before the check if a VR needs to be created is done? A vm is started (on deploy) and at that moment the VR is started if it is not running. I am looking for the possibility of a race condition there. Can it be that the GC doesn't see the nic (yet) and decides to shutdown the VR, while the VM start process has seen the network implemented and does not try to start the VR for it?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@harikrishna-patnala can you approve now?
@weizhouapache any test spec or advice possible?

**Comment by weizhouapache** [comment]
@DaanHoogland @harikrishna-patnala @shwstppr 
I did some testing, look good.

Here the steps
- create a network and deploy/start a new vm
- update the nics_count from 1 to 0 (update op_networks set nics_count =0 where id=295;)
- wait for network GC to run

prior to this PR: network is shutdown  (unexpected as there is a running VM)
```
2023-11-20 07:55:04,616 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Network-Scavenger-1:ctx-c2c336b0) (logid:0a493c1f) Lock is acquired for network Network {"id": 295, "name": "cks-001-network", "uuid": "0306e71a-d587-4437-84a7-3839fd0f53d8", "networkofferingid": 21} as a part of network shutdown
2023-11-20 07:55:04,623 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Network-Scavenger-1:ctx-c2c336b0) (logid:0a493c1f) Sending network shutdown to VirtualRouter
2023-11-20 07:55:04,625 DEBUG [c.c.n.r.VirtualNetworkApplianceManagerImpl] (Network-Scavenger-1:ctx-c2c336b0) (logid:0a493c1f) Stopping router VM instance {"id":223,"instanceName":"r-223-VM","type":"DomainRouter","uuid":"74915ce2-bcf5-4f68-b2f5-89274093d10a"}
```

with this PR, network is not shutdown (expected as there is a running VM)
```
2023-11-20 09:27:50,359 DEBUG [c.c.n.NetworkModelImpl] (Network-Scavenger-1:ctx-8419537c) (logid:8b4be21a) Network id=295 is not ready for GC as it has vms that are not Stopped at the moment
```

stop the vm, and wait for network GC, network is shutdown  (expected as there is not a running VM)
```
2023-11-20 09:37:50,394 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Network-Scavenger-1:ctx-f6ef312a) (logid:4343d2e6) Lock is acquired for network Network {"id": 210, "name": "admin-001", "uuid": "35f01c5c-f543-4c67-93ed-79bbeb52d7e1", "networkofferingid": 10} as a part of network shutdown
2023-11-20 09:37:50,401 DEBUG [o.a.c.e.o.NetworkOrchestrator] (Network-Scavenger-1:ctx-f6ef312a) (logid:4343d2e6) Sending network shutdown to VirtualRouter
2023-11-20 09:37:50,404 DEBUG [c.c.n.r.VirtualNetworkApplianceManagerImpl] (Network-Scavenger-1:ctx-f6ef312a) (logid:4343d2e6) Stopping router VM instance {"id":163,"instanceName":"r-163-VM","type":"DomainRouter","uuid":"6ee36e60-a388-41d6-8688-162b0b402a55"}
```

**Review by shwstppr — APPROVED** [review]
Code change looks okay for an improvement.
I'm not sure if this is a bug or a root cause for an issue as the only way is to update DB manually.

**Comment by weizhouapache** [comment]
> Code change looks okay for an improvement. I'm not sure if this is a bug or a root cause for an issue as the only way is to update DB manually.

@shwstppr @DaanHoogland 
I think there is a way to reproduce the issue but I did not test it yet
- create and start a vm, nics_count is increased by 1 (e.g. 0 -> 1)
- stop the vm, nics_count is decreased by 1 (e.g. 1 -> 0)
- start the vm out-of-band (e.g. in vCenter)
- cloustack detects the VM is running, and changes the state from Stopped to Running, but nics_count is not changed (e.g. still 0)
This is just an example that nics_count is incorrect.

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@weizhouapache @DaanHoogland do we need any additional testing here or is it good to merge?

**Comment by weizhouapache** [comment]
> @weizhouapache @DaanHoogland do we need any additional testing here or is it good to merge?

@shwstppr 
I have manually tested it, it looks good to me. 
However, as you know, I am the author of this PR, please ignore my lgtm  ... :-D

**Comment by DaanHoogland** [comment]
> > @weizhouapache @DaanHoogland do we need any additional testing here or is it good to merge?
> 
> @shwstppr I have manually tested it, it looks good to me. However, as you know, I am the author of this PR, please ignore my lgtm ... :-D

I will repro @weizhouapache last testing description and merge if ok.

**Comment by DaanHoogland** [comment]
> I think there is a way to reproduce the issue but I did not test it yet
> 
>     * create and start a vm, nics_count is increased by 1 (e.g. 0 -> 1)
> 
>     * stop the vm, nics_count is decreased by 1 (e.g. 1 -> 0)
> 
>     * start the vm out-of-band (e.g. in vCenter)
> 
>     * cloustack detects the VM is running, and changes the state from Stopped to Running, but nics_count is not changed (e.g. still 0)
>       This is just an example that nics_count is incorrect.

I reproduced the faulty state this way
- set the `network.gc.interval` and `network.gc.timeout` both to 30 secs
- create and start a vm
- stop the vm
- start the vm out of bounds
pre this PR the network gets shutdown despite the vm being started
with the PR the network keeps running

merging

**Comment by weizhouapache** [comment]
great, thanks a lot for the testing @DaanHoogland  !

**Pull request merged**